Add basic state recovery support

dongahn opened the pull request:
This PR implements the basic state recovery semantics.

Recover the state of both the running job queue and the resource graph data store for those running jobs. Meet the state-recovery requirement for a key near-term use case in which both `resource` and `qmanager` modules are shut down and reloaded in a coordinated fashion. Assume two conditions for full recovery:

-  The RV1 JSON object for each and every currently running job includes the `scheduling` key, therefore the JGF section.
- The `resource` module is reloaded before `qmanager`.

Lay some additional work to prepare for more comprehensive failure recovery in the future. If only the `qmanager` module is shut down and reloaded, still fully recover the state. If both qmanager and resource are reloaded, but the RV1 JSON  of a running job does not include the `scheduling` key, `qmanager` will fail to load because  it can't recover the resource state. 

[Table 1] Current flux-sched state recovery semantics
| resource  | qmanager | JGF  | no-JGF |
|----------|-----------|------|--------|
|    reload   | reload          | yes     | no  |
|    running |  reload        | yes     | yes         |
|    reload    | running      | no      | no.         |

Resolve #470

dongahn commented:
I added the [WIP] tag as I still need to add the queue name of each job into `RV1`. The queue name in `R` will allow `qmanager`'s hello callback to add the running job to the right queue.

dongahn commented:
I just found a new sharpness script (`./t1009-recovery-none.t`) is hung. I have the following test copied from a another test and even w/ that, it appears the whole test is hung on exit.

```bash
# Reload the core scheduler so that rc3 won't hang waiting for
# queue to become idle after jobs are canceled.
test_expect_success 'load sched-simple module' '
    flux module load sched-simple
'
```

```console
ok 6 - removing resource and qmanager modules
2020-06-05T16:39:51.357186Z sched-simple.err[0]: hello: rlist_remove (rank0/core[12-15]): No such file or directory
ok 7 - load sched-simple module
# passed all 7 test(s)
1..7
2020-06-05T16:39:51.473915Z broker.err[0]: rc3: flux-job: Canceled 1 jobs (0 errors)
2020-06-05T16:39:51.499430Z sched-simple.err[0]: free: rank0/core[12-15]: No such file or directory
```

@grondo or @garlick: 
This is a case where one job is still running by the time sched-simple is being loaded. Do you have any clue why this is hung and is the err message is related to the hang at all?

dongahn commented:
BTW, 

> # Reload the core scheduler so that rc3 won't hang waiting for
# queue to become idle after jobs are canceled.

I don't fully understand why `sched-simple` is required, but in production I will assume rc3's requirement can be satisfied by fluxion as well. IOW, `sched-simple` is there only for the purpose of testing.

dongahn commented:
> @grondo or @garlick:
This is a case where one job is still running by the time sched-simple is being loaded. Do you have any clue why this is hung and is the err message is related to the hang at all?

Another piece of info: I also didn't see this hang with my old PR (i.e., before this merging and new flux-core).

garlick commented:
See @grondo's comment in 41e955d7373b0c919d3b42d0187ab0ef1dfccda6

Does a similar thing need to happen here?  (Haven't looked yet in detail)

dongahn commented:
> See @grondo's comment in 41e955d

> Does a similar thing need to happen here? (Haven't looked yet in detail)

Ah... Yes, this does seem like the cause. Let me test this out.

dongahn commented:
41e955d

```
flux module unload sched-simple &&
```

I never used `unload` myself and it doesn't show up in the main help message:

```
Usage: flux-module COMMAND [OPTIONS]
  -h, --help             Display this message.

flux module subcommands:
   list            List loaded modules
   remove          Unload module
   load            Load module
   reload          Reload module
   info            Display module info
   stats           Display stats on module
   debug           Get/set module debug flags
```

Is there a difference between `remove` and `unload` or is this just an alias?

dongahn commented:
OK. This fixes the hang.

```diff
 # Reload the core scheduler so that rc3 won't hang waiting for
 # queue to become idle after jobs are canceled.
 test_expect_success 'load sched-simple module' '
+    flux module reload -f resource &&
     flux module load sched-simple
 '
```

garlick commented:
Unload is just an alias for remove.

dongahn commented:
Travis CI is not happy! Let me now try to quickly add queue name support within RV1: https://github.com/flux-framework/rfc/issues/240.

dongahn commented:
Just added multi queue support for state recovery. Once Travis turns green, I will take off WIP tag.

dongahn commented:
Two Travis configurations failed:

> ./t1010-recovery-multiqueue.t: line 18: syntax error near unexpected token `return'
1847./t1010-recovery-multiqueue.t: line 18: `    [[ $? -ne 0 ]] return $?'

Too much "bash"sm. Will fix.

dongahn commented:
Travis CI is green now. I nominated a few people as reviewers :-). Thank you in advance.

garlick commented:
> But the resources that those running jobs are using can be scheduled to other jobs (i.e., overscheduling) because the resource module won't be able to reconstruct the state.

This would seem like a fundamental failure.  Should state recovery by always enabled? What is the use case for disabling it?

garlick commented:
This is a rather big PR.  Would it make sense to pull out all the cleanup not related to state reconstruction to a separate PR?

dongahn commented:
@garlick: thank you for the quick response.

> This would seem like a fundamental failure. Should state recovery by always enabled? 

To enable this always, we need to make the `scheduling` key in RV1 mandatory. This would be fine for the system instance, but this would add rather high overheads for nested instances in particular for high throughput loads. 

For throughput-oriented scheduling specialization, users may trade an ability to recover the state on reload with performance for their nested instances.

As part of RV2, if the graph can be sufficiently reduced and compressed and thus won't be on the critical path, this can be turned on by default. But until then, I think we should still make this trade off possible.

In general, how important is it to recover the states for single user instance though?

dongahn commented:
> This is a rather big PR. Would it make sense to pull out all the cleanup not related to state reconstruction to a separate PR?

Sure. I can split this into two.

garlick commented:
> For throughput-oriented scheduling specialization, users may trade an ability to recover the state on reload with performance for their nested instances.

That makes sense to me, however 

> but the resources that those running jobs are using can be scheduled to other jobs (i.e., overscheduling) because the resource module won't be able to reconstruct the state.

should the scheduler continue in this case, or give up since allowing resources to be exclusively allocated to multiple jobs is a pretty bad situation?

dongahn commented:
> should the scheduler continue in this case, or give up since allowing resources to be exclusively allocated to multiple jobs is a pretty bad situation?

This is a good point. I actually thought about this and realized either has a pros and cons. 

- If the scheduler cannot be loaded, users cannot utilize idling resources until all running jobs complete.

- If the scheduler is loaded, users can over-schedule jobs. 

For high throughput loads, I slightly preferred overscheduling to underutilization, but perhaps this was an incorrect line of thoughts?

garlick commented:
Personally, I think it's game over if the scheduler doesn't know what's allocated.  Hopefully not a condition that would arise from normal use?

dongahn commented:
> Personally, I think it's game over if the scheduler doesn't know what's allocated. Hopefully not a condition that would arise from normal use?

I also don't have a strong preference one way or the other  because like you said this wouldn't arise often for normal cases. 

I can very easily change that semantics to discontinue. Maybe if such a case occurs, users can load `sched-simple` and continue as well. 

I do hope we can advance RV2 and compression/condensing logic so that this becomes a moot point.

dongahn commented:
@SteVwonder and @milroy: I can split this PR into two and change the failure semantics as discussed above. I'll go ahead and do this first before you start your reviews. Let me know if this doesn't work ASAP.

dongahn commented:
A new PR with misc. cleanup commits is now posted at PR #666. That PR should be reviewed and merged before this PR.

dongahn commented:
@garlick: I looked at the PR description once again and realized I may have misinformed you. 

>> but the resources that those running jobs are using can be scheduled to other jobs (i.e., overscheduling) because the resource module won't be able to reconstruct the state.

> should the scheduler continue in this case, or give up since allowing resources to be exclusively allocated to multiple jobs is a pretty bad situation?

This trade-off is made only when the `resource-recovery-on-load=false` option is given. If `resource-recovery-on-load=true`, which would be likely default, `qmanager` reload indeed aborts when it detects it can't recover the state. 

```console
ahn1@4a292901b590:/usr/src$ flux module load sched-fluxion-qmanager
2020-06-09T04:27:22.887298Z sched-fluxion-resource.err[0]: parse_R: no scheduling key in R
2020-06-09T04:27:22.887342Z sched-fluxion-resource.err[0]: run_update: parsing R: No such file or directory
2020-06-09T04:27:22.887361Z sched-fluxion-resource.err[0]: update_request_cb: update failed (id=297376153600): No such file or directory
2020-06-09T04:27:22.887797Z sched-fluxion-qmanager.err[0]: jobmanager_hello_cb: reconstruct (id=297376153600 queue=default): No such file or directory
2020-06-09T04:27:22.887957Z sched-fluxion-qmanager.err[0]: hello: error loading R for id=297376153600: No such file or directory
2020-06-09T04:27:22.888006Z sched-fluxion-qmanager.err[0]: handshake_jobmanager: schedutil_hello: No such file or directory
2020-06-09T04:27:22.888023Z sched-fluxion-qmanager.err[0]: enforce_options: handshake_jobmanager: No such file or directory
2020-06-09T04:27:22.888031Z sched-fluxion-qmanager.err[0]: mod_start: enforce_queue_policy: No such file or directory
2020-06-09T04:27:22.889825Z sched-fluxion-qmanager.crit[0]: fatal error: No such file or directory
flux-module: cmb.insmod: No such file or directory
```

Would you still suggest we should drop this option for good and disallow a way to override this behavior? Wouldn't it be ok to leave this as a rarely used option for users for a few convoluted cases, as well as testing and debugging, though?  

Like I said above, I still don't have a strong preference one way or the other. But I wanted to let you know this subtlety.

garlick commented:
If I understand, I don't see a downside to dropping the option.  It sounds
like it's use can lead to a bad situation

On Mon, Jun 8, 2020, 9:33 PM Dong H. Ahn <notifications@github.com> wrote:

> @garlick <https://github.com/garlick>: I looked at the PR description
> once again and realized I may have misinformed you.
>
> but the resources that those running jobs are using can be scheduled to
> other jobs (i.e., overscheduling) because the resource module won't be able
> to reconstruct the state.
>
> should the scheduler continue in this case, or give up since allowing
> resources to be exclusively allocated to multiple jobs is a pretty bad
> situation?
>
> This trade-off is made only when the resource-recovery-on-load=false
> option is given. If resource-recovery-on-load=true, which would be likely
> default, qmanager reload indeed aborts when it detects it can't recover
> the state.
>
> ahn1@4a292901b590:/usr/src$ flux module load sched-fluxion-qmanager2020-06-09T04:27:22.887298Z sched-fluxion-resource.err[0]: parse_R: no scheduling key in R2020-06-09T04:27:22.887342Z sched-fluxion-resource.err[0]: run_update: parsing R: No such file or directory2020-06-09T04:27:22.887361Z sched-fluxion-resource.err[0]: update_request_cb: update failed (id=297376153600): No such file or directory2020-06-09T04:27:22.887797Z sched-fluxion-qmanager.err[0]: jobmanager_hello_cb: reconstruct (id=297376153600 queue=default): No such file or directory2020-06-09T04:27:22.887957Z sched-fluxion-qmanager.err[0]: hello: error loading R for id=297376153600: No such file or directory2020-06-09T04:27:22.888006Z sched-fluxion-qmanager.err[0]: handshake_jobmanager: schedutil_hello: No such file or directory2020-06-09T04:27:22.888023Z sched-fluxion-qmanager.err[0]: enforce_options: handshake_jobmanager: No such file or directory2020-06-09T04:27:22.888031Z sched-fluxion-qmanager.err[0]: mod_start: enforce_queue_policy: No such file or directory2020-06-09T04:27:22.889825Z sched-fluxion-qmanager.crit[0]: fatal error: No such file or directoryflux-module: cmb.insmod: No such file or directory
>
> Would you still suggest we should drop this option for good and disallow a
> way to override this behavior? Wouldn't it be ok to leave this as a rarely
> used option for users for a few convoluted cases, as well as testing and
> debugging, though?
>
> Like I said above, I still don't have a strong preference one way or the
> other. But I wanted to let you know this subtlety.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-sched/pull/663#issuecomment-641021769>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPW3DPJ2KVTK6WHYEFE3RVW3RHANCNFSM4NTKEYWQ>
> .
>

SteVwonder commented:
> qmanager reload indeed aborts when it detects it can't recover the state.

I haven't dug into this PR yet, so I suspect this will become clear after I do that, but presumably if there are no jobs in the queue when Fluxion is reloaded, everything will work just fine since there is no job state to reconstruct?  For example, is the following situation possible:

- Fluxion set to use RV1, some jobs submitted, scheduled, and started
- Fluxion unloaded
- `flux queue drain` is run; all jobs are now completed and the queue empty
- Fluxion is reloaded with `resource-recovery-on-load=true`
- more jobs submitted, scheduled, and started

dongahn commented:
@SteVwonder: that is correct. But we will no longer have the `resource-recovery-on-load=` option though.

dongahn commented:
OK. I rebased this PR with the current master with PR #666, dropped `resource-recovery-on-load` option and adjusted tests and such. The CI is green so I will open this for further reviews.

SteVwonder commented on qmanager/modules/qmanager.cpp:
```diff
@@ -100,7 +102,9 @@ static int process_config_file (std::shared_ptr<qmanager_ctx_t> &ctx)
     json_object_foreach (conf, k, v) {
-        std::string value = json_string_value (v);
+        std::string value = json_dumps (v, JSON_ENCODE_ANY|JSON_COMPACT);
+        if (json_typeof (v) == JSON_STRING)
+            value = value.substr (1, value.length () - 2);
```
I assume this is here to strip the quotes off the ends of the string?  For some reason this feels like it could be fragile, I'm not sure though.  Any opposition to just grabbing the string value via the jansson API?

```suggestion
        if (json_typeof (v) == JSON_STRING)
            value = json_string_value (v);
```

If you want to keep the `substr`, can you leave a short comment as to the purpose?

dongahn replied:
That was my initial code but I changed it so that one can add non string values within `sched-fluxion-qmanager.toml` like our dropped support for `resource-recovery-on-load`

```
resource-recovery-on-load = true
```

As opposed to 

```
resource-recovery-on-load = "true"
```

can be supported.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -653,6 +653,110 @@ static int track_schedule_info (std::shared_ptr<resource_ctx_t> &ctx,
+    }
+    jgf = jgf_str;
+    free (jgf_str);
+    starttime = static_cast<uint64_t> (st);
```
Sorry this is a bit convoluted. But `int64_t` is used to make error checking a bit handy (i.e., an error if -1) for "starting" scheduled points. A starting scheduled point cannot exceed `MAX_INT64`. If it did, it should be caught at the producer side (`json_pack`) side. So assigning `st` back to `starttime` should be safe. 

However, this code can be improved by making `st` and `et` `uint64_t` (and casting them accordingly when the are assigned `starttime` and `duration`). Will make the change.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -653,6 +653,110 @@ static int track_schedule_info (std::shared_ptr<resource_ctx_t> &ctx,
+    return rc;
+}
+
+static int R_equal (const std::shared_ptr<resource_ctx_t> &ctx,
```
Agreed. Will change.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -825,6 +853,112 @@ static int run_match (std::shared_ptr<resource_ctx_t> &ctx, int64_t jobid,
+            goto error;
+        }
+        // If a jobid with matching R exists, no need to update
+        ov = ctx->jobs[jobid]->overhead;
```
Yes, good point.  I think getting the lapse time for `R_equal` and assigning that to `ov` will do in this case.

BTW, the `overhead` key of the response isn't really used by the remote caller when used for real use cases (state reconstruction): `qmanager` won't send a response (or a new request) to `job-manager` when the state is recovered from within `hello` callback. Also, I wouldn't change the overhead record (`ctx->jobs[jobid]->overhead`). 

But responding with ~0 overhead does make sense to me for `flux ion-resource` (or `flux ion` later).

dongahn commented on t/scripts/flux-ion-resource.py:
```diff
@@ -111,6 +116,20 @@ def match_reserve_action (args):
+"""
+def update_action (args):
+    with open (args.RV1, 'r') as stream:
+        RV1 = json.dumps (json.loads (stream.read ()))
```
Good to know. Thanks.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -653,6 +653,110 @@ static int track_schedule_info (std::shared_ptr<resource_ctx_t> &ctx,
+    duration = static_cast<uint64_t> (et) - static_cast<uint64_t> (st);
+
+freemem_out:
+    json_decref (o);
```
Ah... good call. This is an error path and we want to preserved `errno`!

Just out of curiosity: do you know if `json_decref` sets `errno`? Apparently, I am not a big fan of using a global variable to propagate errors, but we have no other alternatives... :-)

garlick replied:
In flux-core, we've been operating under the defensive assumption that `free()`, and hence destructors that internally call `free()`, and don't _preserve_ errno,  _could_ set errno.  

https://stackoverflow.com/questions/30569981/does-free-set-errno/30571921#30571921

dongahn replied:
Great info. Thanks @garlick. 

Just to make sure I am on the same page on the general errno propagation semantics. Looking at flux-core code, it seems like the semantics is best effort:

1. Capture `errno` on the first failure detected
2. Don't reset `errno` on the cleanup/exit paths of this failure where it makes sense

W/ respect to 2) from flux-core, I see two styles: A) only preserve `errno` possibly set by non-Flux function calls (e.g., jansson calls, free...); B) preserve the `errno` even for *all* calls including Flux API calls themselves.

I can see an argument for A vs. B., but it feels like having a general semantics written down for consistency could be beneficial. If you folks already discussed this, my apology. -- Seems like the semantics can go in a RFC -- I don't know if something like this can go into the coding style RFC though.

garlick replied:
I'm not sure I understand your A vs B.

A function that purports to set errno on failure may set it to any value it likes.  Usually you want that to reflect the root (first) cause of a problem but not always.

Jansson makes no claims about setting errno on failure.  But it probably does _sometimes_ given that it must make calls to `malloc()` for example.

The main thing is (a) if your contract with users is that errno will be set on error, you must make sure it is not set to zero on error, and (b) once you have errno set the way you want it you must avoid clobbering it since it is a stupid global variable.

Destructors are often called after errno is set, and destructors are usually void-returning functions that ignore errors.  Thus if they call any functions that set errno and then ignore the failures of those functions, they may clobber your carefully set errno value.

(Sorry if I'm just stating obvious stuff because I didn't quite get your question!)

BTW I don't think it's a forgone conclusion that fluxion internals must use this same error handling style as flux-core.  Obviously they have to deal with flux-core's error semantics but they could adopt something different for how those errors are propagated.

dongahn replied:
> Destructors are often called after errno is set, and destructors are usually void-returning functions that ignore errors. Thus if they call any functions that set errno and then ignore the failures of those functions, they may clobber your carefully set errno value.

Yes, this was essentially my question -- I have seen a number of examples where a destructor in the Flux API set is called on error paths without no errno wrap. My guess the Flux destructors themselves are written to preserve errno within itself so this clobbering won't happen.

garlick replied:
Correct!

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -825,6 +853,112 @@ static int run_match (std::shared_ptr<resource_ctx_t> &ctx, int64_t jobid,
+    std::string jgf;
+    std::string R2;
+
+    if ( (rc = gettimeofday (&start, NULL)) < 0) {
+        flux_log_error (ctx->h, "%s: gettimeofday", __FUNCTION__);
+        goto done;
+    }
+    if ( (rc = parse_R (ctx, R, jgf, at, duration)) < 0) {
+        flux_log_error (ctx->h, "%s: parsing R", __FUNCTION__);
+        goto done;
+    }
+    if ( (rc = run (ctx, jobid, jgf, at, duration)) < 0) {
+        flux_log_error (ctx->h, "%s: run", __FUNCTION__);
+        goto done;
+    }
+    if ( (rc = ctx->writers->emit (o)) < 0) {
+        flux_log_error (ctx->h, "%s: writers->emit", __FUNCTION__);
+        goto done;
+    }
+    if ( (rc = gettimeofday (&end, NULL)) < 0) {
+        flux_log_error (ctx->h, "%s: gettimeofday", __FUNCTION__);
+        goto done;
+    }
+    ov = get_elapse_time (start, end);
```
Yes, these should all set `errno` including `gettimeofday()`. 

I also have https://github.com/flux-framework/flux-sched/issues/618 to do a PR only on this topic to tighten up any loosen end. Finally, merging @trws's exception PR should make `sched-fluxion-resource` more robust when a C++ exception arises from within a request handlers.

dongahn commented on resource/hlapi/bindings/c++/reapi_module_impl.hpp:
```diff
@@ -84,6 +84,54 @@ int reapi_module_t::match_allocate (void *h, bool orelse_reserve,
+                                           "overhead", &overhead,
+                                           "R", &rset,
+                                           "at", &scheduled_at)) < 0)
+        goto out;
```
Doh! Good catch.

dongahn commented on t/t1007-recovery-full.t:
```diff
@@ -0,0 +1,150 @@
+test_expect_success 'recovery: loading flux-sched modules works (rv1)' '
+    flux module remove sched-simple &&
+    flux module reload -f resource &&
+    load_resource load-whitelist=node,core,gpu match-format=rv1 &&
```
Will do this as part of a future PR.

dongahn commented on resource/writers/match_writers.cpp:
```diff
@@ -627,12 +627,46 @@ bool rv1_match_writers_t::empty ()
+            goto ret;
+        }
+        if ((rc = json_object_set_new (attrs, kv.first.c_str (), stro)) == -1) {
+            json_decref (attrs);
```
I thought we determined free memory in this case is unsafe.

https://github.com/flux-framework/flux-core/issues/2803#issuecomment-595617111

If we have other places where `stro` equivalents are freed, we need to change it; double free can lead to a broker crash.

This may have changed in a more recent jansson library. @chu11?

garlick replied:
Oof!  Thanks for the reminder!  I'd forgotten this affects functions other than `json_pack()`.  It's unpleasant to have to choose between a double free and memory leak.

One way around it is to avoid `json_pack()` with `o` or any of the `*_new()` "object stealing" interfaces in jansson.  It means a bit of extra reference count management in the non-error case.

dongahn commented:
@garlick and @SteVwonder: I added fixup commits and I believe I addressed the issues captured in your review comments. Let me know if there is anything else, you want me to do before this is merged.

SteVwonder commented:
@dongahn: feel free to squash the commit and set MWP whenever you are ready

dongahn commented:
Thanks!
